instrumentation/rxjava-2.0/library/src/main/java/io/opentelemetry/instrumentation/rxjava2/TracingAssembly.java

```diff
@@ -0,0 +1,338 @@
+  @GuardedBy("TracingAssembly.class")
+  private static Function<? super ConnectableObservable, ? extends ConnectableObservable>
+      oldOnConnectableObservableAssembly;
+
+  @SuppressWarnings("rawtypes")
+  @GuardedBy("TracingAssembly.class")
+  private static BiFunction<
+          ? super Completable, ? super CompletableObserver, ? extends CompletableObserver>
+      oldOnCompletableSubscribe;
+
+  @SuppressWarnings("rawtypes")
+  @GuardedBy("TracingAssembly.class")
+  private static BiFunction<? super Single, ? super SingleObserver, ? extends SingleObserver>
+      oldOnSingleSubscribe;
+
+  @SuppressWarnings("rawtypes")
+  @GuardedBy("TracingAssembly.class")
+  private static BiFunction<? super Maybe, ? super MaybeObserver, ? extends MaybeObserver>
+      oldOnMaybeSubscribe;
+
+  @SuppressWarnings("rawtypes")
+  @GuardedBy("TracingAssembly.class")
+  private static BiFunction<? super Flowable, ? super Subscriber, ? extends Subscriber>
+      oldOnFlowableSubscribe;
+
+  @SuppressWarnings("rawtypes")
+  @GuardedBy("TracingAssembly.class")
+  private static Function<? super ConnectableFlowable, ? extends ConnectableFlowable>
+      oldOnConnectableFlowableAssembly;
+
+  @SuppressWarnings("rawtypes")
+  @GuardedBy("TracingAssembly.class")
+  private static Function<? super ParallelFlowable, ? extends ParallelFlowable>
+      oldOnParallelAssembly;
+
+  @GuardedBy("TracingAssembly.class")
+  private static boolean enabled;
+
+  private TracingAssembly() {}
+
+  public static synchronized void enable() {
+    if (enabled) {
+      return;
+    }
+
+    enableObservable();
+
+    enableConnectableObservable();
+
+    enableCompletable();
+
+    enableSingle();
+
+    enableMaybe();
+
+    enableFlowable();
+
+    enableConnectableFlowable();
+
+    enableParallel();
+
+    enabled = true;
+  }
+
+  public static synchronized void disable() {
+    if (!enabled) {
+      return;
+    }
+
+    disableSingle();
+
+    disableMaybe();
+
+    disableFlowable();
+
+    disableParallel();
+
+    disableObservable();
+
+    disableConnectableObservable();
+
+    disableCompletable();
+
+    disableConnectableFlowable();
+
+    enabled = false;
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private static void enableParallel() {
+    oldOnParallelAssembly = RxJavaPlugins.getOnParallelAssembly();
+    RxJavaPlugins.setOnParallelAssembly(
+        compose(
+            oldOnParallelAssembly,
+            new ConditionalOnCurrentContextFunction<ParallelFlowable>() {
+              @Override
+              ParallelFlowable applyActual(ParallelFlowable parallelFlowable, Context ctx) {
+                return new TracingParallelFlowable(parallelFlowable, ctx);
+              }
+            }));
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private static void enableConnectableFlowable() {
+    oldOnConnectableFlowableAssembly = RxJavaPlugins.getOnConnectableFlowableAssembly();
+    RxJavaPlugins.setOnConnectableFlowableAssembly(
```

Comment:
I'm curious why keep setOnConnectableXxxAssembly? 
Because the class uses setOnXxxSubscribe instead of using setOnXxxAssembly for using context when subscribing.

Reply:
I will try to remove this and check if tests pass.

Reply:
You are correct, I was able to remove them.

Reply:
`onParallelAssemblly` also? And looks like RxJava2 didn't provide hook for subscribe yet. 
https://github.com/ReactiveX/RxJava/issues/7190

Reply:
Looks like they'll add the missing plugin method in 3.x: https://github.com/ReactiveX/RxJava/pull/7191/files